zebra-consensus/src/verify.rs

```diff
@@ -0,0 +1,60 @@
+//! Context-dependent block and transaction verification for Zebra.
+//!
+//! Verification occurs in two stages:
+//!   - context-free verification of signatures, proofs, and scripts, handled by `zebra-chain` and
+//!     `zebra-script`
+//!   - context-dependent verification of the chain, handled by `zebra-consensus` (this module)
+//!
+//! Verification is provided via a `tower::Service`, to support batching.
+
+use std::sync::Arc;
+use zebra_chain::block::Block;
+use zebra_chain::transaction::Transaction;
+
+/// A context-dependent block or transaction verification request.
+///
+/// This has a `Flush` variant, to allow manually flushing queued verification
+/// requests, even when the batching service is wrapped in other `tower` layers.
+#[derive(Debug)]
+// The associated data doesn't need separate documentation.
+#[allow(missing_docs)]
+pub enum Request {
+    /// Verify the block and its transactions, and check them against the chain state
+    VerifyBlock { block: Arc<Block> },
+    /// Verify the transaction, and check it against the chain state
+    VerifyTransaction { transaction: Arc<Transaction> },
+    /// Flush incomplete verification batches
+    Flush,
+}
+
+/// A context-dependent block or transaction verification response.
+///
+/// First, the signatures, proofs, and scripts in each block and transaction are verified. If this
+/// context-free verification succeeds, then the block or transaction is checked against the current
+/// chain state. If the block or transaction is valid in the chain, it is added to the relevant part
+/// of the `zebra-state`. The future returns an `Added` response.
+///
+/// If context-free verification succeeds, but context-dependent verification is pending chain state
+/// updates, then the future returns a `Pending` response. This response describes the type of state
```

Comment:
This seems strange to me. I would expect a future that is waiting for more information to return a `Poll::pending`, not a `Poll::Ready(Response::PendingSomething)`. It seems like this is creating a future like interface on top of a service where we're using multiple request / responses to get the state of the _actual_ future we're representing behind the scenes.

This is fine for a first pass stub I think but I feel like we should review this more carefully to be sure this is the best approach, or at least so I can understand why this is the best approach.

Reply:
Thanks, I asked some questions in slack that will help me refine this design.